core/src/main/java/org/apache/calcite/rex/RexLiteral.java

```diff
@@ -568,133 +592,149 @@ public void printAsJava(PrintWriter pw) {
+      sb.append(value.toString());
+      break;
+    case DOUBLE:
+      assert value instanceof BigDecimal;
+      sb.append(Util.toScientificNotation((BigDecimal) value));
+      break;
+    case BIGINT:
+      assert value instanceof BigDecimal;
+      long narrowLong = ((BigDecimal) value).longValue();
+      sb.append(String.valueOf(narrowLong));
+      sb.append('L');
+      break;
+    case BINARY:
+      assert value instanceof ByteString;
+      sb.append("X'");
+      sb.append(((ByteString) value).toString(16));
+      sb.append("'");
+      break;
+    case NULL:
+      assert value == null;
+      sb.append("null");
+      break;
+    case SARG:
+      assert value instanceof Sarg;
+      //noinspection unchecked,rawtypes
+      Util.asStringBuilder(sb, sb2 ->
+          printSarg(sb2, (Sarg) value, type));
+      break;
+    case SYMBOL:
+      assert value instanceof Enum;
+      sb.append("FLAG(");
+      sb.append(value.toString());
+      sb.append(")");
+      break;
+    case DATE:
+      assert value instanceof DateString;
+      sb.append(value.toString());
+      break;
+    case TIME:
+    case TIME_WITH_LOCAL_TIME_ZONE:
+      assert value instanceof TimeString;
+      sb.append(value.toString());
+      break;
+    case TIMESTAMP:
+    case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+      assert value instanceof TimestampString;
+      sb.append(value.toString());
+      break;
+    case INTERVAL_YEAR:
+    case INTERVAL_YEAR_MONTH:
+    case INTERVAL_MONTH:
+    case INTERVAL_DAY:
+    case INTERVAL_DAY_HOUR:
+    case INTERVAL_DAY_MINUTE:
+    case INTERVAL_DAY_SECOND:
+    case INTERVAL_HOUR:
+    case INTERVAL_HOUR_MINUTE:
+    case INTERVAL_HOUR_SECOND:
+    case INTERVAL_MINUTE:
+    case INTERVAL_MINUTE_SECOND:
+    case INTERVAL_SECOND:
+      assert value instanceof BigDecimal;
+      sb.append(value.toString());
+      break;
+    case MULTISET:
+    case ROW:
+      final List<RexLiteral> list = (List) value;
+      Util.asStringBuilder(sb, sb2 ->
+          Util.printList(sb, list.size(), (sb3, i) ->
+              sb3.append(list.get(i).computeDigest(includeType))));
+      break;
+    case GEOMETRY:
+      final String wkt = GeoFunctions.ST_AsWKT((Geometries.Geom) value);
+      sb.append(wkt);
+      break;
+    default:
+      assert valueMatchesType(value, typeName, true);
+      throw Util.needToImplement(typeName);
+    }
+  }
+
+  private static <C extends Comparable<C>> void printSarg(StringBuilder sb,
+      Sarg<C> sarg, RelDataType type) {
+    sarg.printTo(sb, (sb2, value) ->
+        sb2.append(toLiteral(type, value)));
+  }
+
+  /** Converts a value to a temporary literal, for the purposes of generating a
+   * digest. Literals of type ROW and MULTISET require that their components are
+   * also literals. */
+  private static RexLiteral toLiteral(RelDataType type, Comparable<?> value) {
+    final SqlTypeName typeName = strictTypeName(type);
+    switch (typeName) {
```

Comment:
Seems an useless call of `strictTypeName(type)`.

Reply:
Remember that `RexLiteral.typeName` is not always equal to `RexLiteral.type.getTypeName()`. It holds the type of literal (e.g. SARG) when the value is of a different type.

So, try removing that 'useless call' and tests will fail.